src/x86.rs

```diff
@@ -132,222 +132,228 @@ impl<V: Vector> VectorHash<V> {
+    needle: N,
+}
 
-        impl $name {
-            /// Creates a new searcher for `needle`. By default, `position` is
-            /// set to the last character in the needle.
-            #[target_feature(enable = "avx2")]
-            pub unsafe fn new(needle: Box<[u8]>) -> Self {
-                let position = needle.len() - 1;
-                Self::with_position(needle, position)
-            }
+impl<N: Needle> Avx2Searcher<N> {
+    /// Creates a new searcher for `needle`. By default, `position` is
+    /// set to the last character in the needle.
+    #[target_feature(enable = "avx2")]
+    pub unsafe fn new(needle: N) -> Self {
+        let position = needle.as_ref().len() - 1;
+        Self::with_position(needle, position)
+    }
 
-            /// Same as `new` but allows additionally specifying the `position`
-            /// to use.
-            #[target_feature(enable = "avx2")]
-            pub unsafe fn with_position(needle: Box<[u8]>, position: usize) -> Self {
-                assert!(!needle.is_empty());
-                assert!(position < needle.len());
-
-                let scalar_hash = ScalarHash::from(needle.as_ref());
-                let sse2_hash = VectorHash::new(needle[0], needle[position]);
-                let avx2_hash = VectorHash::new(needle[0], needle[position]);
-
-                Self {
-                    needle,
-                    position,
-                    scalar_hash,
-                    sse2_hash,
-                    avx2_hash,
-                }
-            }
+    /// Same as `new` but allows additionally specifying the `position`
+    /// to use.
+    #[target_feature(enable = "avx2")]
+    pub unsafe fn with_position(needle: N, position: usize) -> Self {
+        let slice = needle.as_ref();
+        assert!(!slice.is_empty());
+        assert!(position < slice.len());
+        if let Some(size) = N::SIZE {
+            assert_eq!(size, slice.len());
+        }
 
-            #[inline]
-            fn size(&self) -> usize {
-                if $size > 0 {
-                    $size
-                } else {
-                    self.needle.len()
-                }
-            }
+        let scalar_hash = ScalarHash::from(needle.as_ref());
+        let sse2_hash = VectorHash::new(slice[0], slice[position]);
+        let avx2_hash = VectorHash::new(slice[0], slice[position]);
 
-            #[inline]
-            fn scalar_search_in(&self, haystack: &[u8]) -> bool {
-                debug_assert!(haystack.len() >= self.size());
+        Self {
+            needle,
+            position,
+            scalar_hash,
+            sse2_hash,
+            avx2_hash,
+        }
+    }
 
-                let mut end = self.size() - 1;
-                let mut hash = ScalarHash::from(&haystack[..end]);
+    #[inline]
+    fn size(&self) -> usize {
+        if let Some(size) = N::SIZE {
+            size
+        } else {
+            self.needle.as_ref().len()
+        }
+    }
```

Comment:
Perhaps this method might be better suited in the `Needle` trait? That would allow it to also be used in `new()` and `with_position()`.

Reply:
So I think that would be nice but I am worried people would try to redefine that method in their own implementation of Needle and I am not sure if it's good or bad to let them do that? I would want a "hidden" trait method not callable outside of the crate, maybe using a sealed trait?

Reply:
Ah yeah, good point. In the futures crate, they have a [`FutureExt`](https://docs.rs/futures/0.3.13/futures/future/trait.FutureExt.html) trait which has an implementation for anything that implements the `Future` trait. Maybe in our case we could create a `NeedleExt` trait which is private.

Reply:
Alternatively, I guess an even simpler solution is just a private function `fn size<N: Needle>(needle: &N) -> usize`.

Reply:
I have chosen to use a private helper trait for this and it seems to be working quite well.